### PR TITLE
Refactor stats

### PR DESCRIPTION
**Pull request by StuAA78**

### CODE REVIEW

**Comment by disconnect78** [comment]
Have refactored code as discussed with @Cruikshanks and added the stats for percentage change from previous week.

I ran out of time to do everything I wanted so please note the following:

* The base stat json methid isn't currently used -- the intention was that stats formatter would create an array of stats objects and call the json method of each one (inherited from base_stat) to assemble the output but I couldn't get it to reliably work. There is a base_stat_spec.rb file but the tests are commented out -- not sure if it's better to just delete it for now?
* The :was_renewed trait stopped working! It was previously `{ build(:registration) }` but this no longer works for some reason -- manually adding a dummy value by amending it to `{ "123" }` works but I guess it would be best to investigate and fix this?
* The change stat isn't tested at all. I had it tested prior to the refactor but ran out of time to do the necessary work to test it afterwards.

**Comment by StuAA78** [comment]
codeclimate/total-coverage has failed due to coverage of `base_stat.rb`. One of the issues here is that `raise NotImplementedError` is not being tested in the `calculate_stat` method, however I believe there is precedent for this not being tested?

**Comment by Cruikshanks** [comment]
That's right @StuAA78 . We have used this pattern in a number of places, but made the conscious decision to not worry about covering it with a Unit test as its more a guide to any future dev building with the base class than some functionality we actually rely on.

You can see examples where we have used it (and our lack of coverage) in the following places

- https://github.com/DEFRA/waste-exemptions-engine/blob/master/app/forms/waste_exemptions_engine/postcode_form.rb#L57
- https://github.com/DEFRA/waste-carriers-engine/blob/master/app/validators/waste_carriers_engine/person_validator.rb#L59
- https://github.com/DEFRA/waste-carriers-engine/blob/master/app/forms/waste_carriers_engine/person_form.rb#L111

File: db/schema.rb [hunk 1]
```diff
@@ -11,22 +11,11 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 20190912112625) do
+ActiveRecord::Schema.define(version: 20190805152649) do
```
Comment (Cruikshanks) [inline db/schema.rb]: Looks like a schema update snuck into your PR. Could you drop this change from the PR to be in the safe side?

File: spec/factories/registration.rb [hunk 2]
```diff
@@ -15,7 +15,8 @@
     registration_exemptions { build_list(:registration_exemption, 3) }
 
     trait :was_renewed do
-      referring_registration { build(:registration) }
+      # referring_registration_id { build(:registration) }
```
Comment (Cruikshanks) [inline spec/factories/registration.rb]: I know you said something about needing to do this to get the test to pass. Could you add a `TODO: ` comment to that effect here e.g.

```ruby
# TODO: Ideally we'd like the referring registration to actually be a
# registration, but this seems to break the tests. We need to investigate why
# but we are happy this is valid as an interim measure.
# referring_registration_id { build(:registration) }
referring_registration_id { "123" }
```

File: spec/models/stats/base_stat_spec.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+module Stats
+  RSpec.describe BaseStat, type: :model do
+    subject { described_class.new(current_registrations, previous_registrations) }
+
+    # context "#json" do
```
Comment (Cruikshanks) [inline spec/models/stats/base_stat_spec.rb]: Again, just a `# TODO: ` comment here to remind us that this test needs to be completed at some point would be useful.

File: db/schema.rb (hunk 1 above)
Comment (StuAA78) [reply db/schema.rb]: It had to happen one last time didn't it! I've reverted the change now.

File: spec/factories/registration.rb (hunk 2 above)
Comment (StuAA78) [reply spec/factories/registration.rb]: Comment added as requested.

File: spec/models/stats/base_stat_spec.rb (hunk 3 above)
Comment (StuAA78) [reply spec/models/stats/base_stat_spec.rb]: Comment added as requested.

**Comment by StuAA78** [comment]
😢

**Pull request merged**